processing/src/main/java/org/apache/druid/frame/field/NumericArrayFieldWriter.java

```diff
@@ -0,0 +1,245 @@
+  {
+    return new NumericArrayFieldWriter(selector, FloatFieldWriter::forArray);
+  }
+
+  /**
+   * Returns the writer for ARRAY<DOUBLE>
+   */
+  public static NumericArrayFieldWriter getDoubleArrayFieldWriter(final ColumnValueSelector selector)
+  {
+    return new NumericArrayFieldWriter(selector, DoubleFieldWriter::forArray);
+  }
+
+  public NumericArrayFieldWriter(final ColumnValueSelector selector, NumericFieldWriterFactory writerFactory)
+  {
+    this.selector = selector;
+    this.writerFactory = writerFactory;
+  }
+
+  @Override
+  public long writeTo(WritableMemory memory, long position, long maxSize)
+  {
+    Object row = selector.getObject();
+    if (row == null) {
+      int requiredSize = Byte.BYTES;
+      if (requiredSize > maxSize) {
+        return -1;
+      }
+      memory.putByte(position, NULL_ROW);
+      return requiredSize;
+    } else {
+
+      List<? extends Number> list = FrameWriterUtils.getNumericArrayFromNumericArray(row);
+
+      if (list == null) {
+        int requiredSize = Byte.BYTES;
+        if (requiredSize > maxSize) {
+          return -1;
+        }
+        memory.putByte(position, NULL_ROW);
+        return requiredSize;
+      }
+
+      // Create a columnValueSelector to write the individual elements re-using the NumericFieldWriter
+      AtomicInteger index = new AtomicInteger(0);
+      ColumnValueSelector<Number> columnValueSelector = new ColumnValueSelector<Number>()
+      {
+        @Override
+        public double getDouble()
+        {
+          final Number n = getObject();
+          assert NullHandling.replaceWithDefault() || n != null;
+          return n != null ? n.doubleValue() : 0d;
+        }
+
+        @Override
+        public float getFloat()
+        {
+          final Number n = getObject();
+          assert NullHandling.replaceWithDefault() || n != null;
+          return n != null ? n.floatValue() : 0f;
+        }
+
+        @Override
+        public long getLong()
+        {
+          final Number n = getObject();
+          assert NullHandling.replaceWithDefault() || n != null;
+          return n != null ? n.longValue() : 0L;
+        }
+
+        @Override
+        public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+        {
+
+        }
+
+        @Override
+        public boolean isNull()
+        {
+          // Arrays preserve the individual element's nullity when they are written and read.
+          // Therefore, when working with SQL incompatible mode, [7, null] won't change to [7, 0] when written to and
+          // read from the underlying serialization (as compared with the primitives). Therefore,
+          // even when NullHandling.replaceWithDefault() is true we need to write null as is, and not convert it to their
+          // default value when writing the array. Therefore, the check is `getObject() == null` ignoring the value of
+          // `NullHandling.replaceWithDefaul()`.
+          return getObject() == null;
```

Comment:
@clintropolis Could you please review this.

Reply:
`NullHandling.replaceWithDefault` is dead to me, but this is true that array columns and expressions do not coerce elements to zeros.

We could also consider only supporting arrays in sql compatible mode :p